### PR TITLE
refactor!: bump typescript-eslint to v8

### PR DESCRIPTION
**Pull request by SukkaW**
https://github.com/typescript-eslint/typescript-eslint/issues/9501

This also bumps the minimum required Node.js version from 16 to 18.18.

### CODE REVIEW

File: test/rules/no-unused-modules.spec.ts [hunk 1]
```diff
@@ -1,15 +1,17 @@
...
+import { version } from 'eslint/package.json'
+import semver from 'semver'
 
 import { test, testFilePath, parsers } from '../utils'
 
 import jsxConfig from 'eslint-plugin-import-x/config/react'
 import typescriptConfig from 'eslint-plugin-import-x/config/typescript'
 import rule from 'eslint-plugin-import-x/rules/no-unused-modules'
 
+const FlatRuleTester = semver.satisfies(version, '8') ? require('eslint/use-at-your-own-risk').FlatRuleTester : require('eslint').RuleTester
```
Comment (SukkaW) [inline test/rules/no-unused-modules.spec.ts:13]: @JoshuaKGoldberg We want to test this specific rule under both flat rule tester and legacy rule tester. And we want the `eslint-plugin-import-x` to be compatible with both ESLint 8.56 and 9.x. Would you like to share your thoughts on how to do this?

File: patches/@typescript-eslint+utils+8.0.0-alpha.41.patch [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+--- a/node_modules/@typescript-eslint/utils/dist/ts-eslint/RuleTester.d.ts
++++ b/node_modules/@typescript-eslint/utils/dist/ts-eslint/RuleTester.d.ts
+@@ -113,7 +113,7 @@ interface RunTests<MessageIds extends string, Options extends readonly unknown[]
+     readonly invalid: readonly InvalidTestCase<MessageIds, Options>[];
+ }
+ interface RuleTesterConfig extends ClassicConfig.Config {
+-    readonly parser: string;
++    readonly parser?: string;
+     readonly parserOptions?: Readonly<ParserOptions>;
+ }
```
Comment (SukkaW) [inline patches/@typescript-eslint+utils+8.0.0-alpha.41.patch:12]: @JoshuaKGoldberg Is the `parser` really required here? It seems that `typescript-eslint` will use the `@typecript-eslint/parser` from the peer dependency (if not explicitly installed) by default.

Comment (JoshuaKGoldberg) [reply patches/@typescript-eslint+utils+8.0.0-alpha.41.patch:12]: Hey good timing, it's not! I _just_ released an update to the v8 beta that switches the `RuleTestBeta` that makes the `parser` optional: https://github.com/typescript-eslint/typescript-eslint/pull/9603.

I'll note that that's actually a breaking change to switch the tester from legacy config to flat config, so the tests in this plugin will need to:

* Move any custom `parser` or `parserOptions` inside a `languageOptions`
* Switch from `require.resolve('@babel/eslint-parser')` to `require('@babel/eslint-parser')`

It should be a manual find-and-replace. The bright side is, you're right that the typescript-eslint parser is now the default.

File: test/rules/no-unused-modules.spec.ts (hunk 1 above)
Comment (JoshuaKGoldberg) [reply test/rules/no-unused-modules.spec.ts:13]: Good question... Both ESLint v9 (to my knowledge) and typescript-eslint v8 (as of today) have fully dropped the old rule tester. So there's no way that I know to get legacy config based testing under the current versions.

Maybe you could install both, like having `"eslint8": "eslint@8.57.0"` in `package.json` `devDependencies` or something like that? I haven't tried that out, just a guess.

You might want to file issue(s) on ESLint and/or typescript-eslint to also export a class like `LegacyRuleTester`. If this ends up being a common ecosystem pain, at least in typescript-eslint we'd want to know & have a place for people to post.

File: patches/@typescript-eslint+utils+8.0.0-alpha.41.patch (hunk 2 above)
Comment (JoshuaKGoldberg) [reply patches/@typescript-eslint+utils+8.0.0-alpha.41.patch:12]: https://deploy-preview-9627--typescript-eslint.netlify.app/blog/announcing-typescript-eslint-v8-beta#flat-configuration-ruletester

File: test/rules/no-unused-modules.spec.ts (hunk 1 above)
Comment (SukkaW) [reply test/rules/no-unused-modules.spec.ts:13]: > Maybe you could install both, like having `"eslint8": "eslint@8.57.0"` in `package.json` `devDependencies` or something like that?

That's a good idea! I could try that out before submitting issues.

**Comment by jerome-benoit** [comment]
Are you considering do an alpha/beta release with TSES v8 support?

**Comment by RobinTail** [comment]
v8 stable is published.
Need this one badly, @SukkaW

**Comment by Mister-Hope** [comment]
```
 WARN  Issues with peer dependencies found
.
└─┬ eslint-plugin-import-x 3.1.0
  └─┬ @typescript-eslint/utils 7.18.0
    └── ✕ unmet peer eslint@^8.56.0: found 9.8.0
```

Now with typescript-eslint v8 warnings are trigged.

**Comment by SukkaW** [comment]
I have been very busy recently. I'm not a full-time open-source contributor and my schedule is packed. The release of typescript-eslint v8 introduced a plethora of breaking changes for eslint plugins and migration takes time.

Currently, `eslint-plugin-import-x` is compatible with typescript-eslint v8. This means that I can leave this PR for another month and revisit then, it still won't break anyone.

**Comment by SukkaW** [comment]
I've decided to postpone the work on this PR. The sheer volume of configurations that need to be migrated without a codemod is overwhelming. I'll focus on backporting https://github.com/import-js/eslint-plugin-import/pull/3018 to `eslint-plugin-import-x` instead, which seems to be a more worthwhile task at the moment.

If anyone would like to take over the PR, please fork my forked repository at https://github.com/SukkaW/eslint-plugin-import-x/tree/tseslint-v8, make changes on the `tseslint-v8` branch, and then submit PRs to my forked repository.

**Comment by controversial** [comment]
Thanks for your hard work on this!!

Could you say more about what the configurations are that need to be migrated? Are you talking about the exported configs from [`src/config`](https://github.com/un-ts/eslint-plugin-import-x/tree/master/src/config)?

**Comment by SukkaW** [comment]
> Thanks for your hard work on this!!
> 
> Could you say more about what the configurations are that need to be migrated? Are you talking about the exported configs from [`src/config`](https://github.com/un-ts/eslint-plugin-import-x/tree/master/src/config?rgh-link-date=2024-08-07T16%3A38%3A01Z)?

The most important thing now is fixing tests. We can only proceed when all tests are passed.

**Comment by merrywhether** [comment]
@SukkaW I put up https://github.com/SukkaW/eslint-plugin-import-x/pull/3 to tackle 88 more tests, bringing the outstanding failures to 33 (about 4 or 5 files left). Hopefully that helps!

**Comment by SukkaW** [comment]
@merrywhether Thanks! I've added a few reviews.

**Comment by SukkaW** [comment]
Thanks for @merrywhether's help, along with my another 2 commits, we now only have 17 failed test cases.

**Comment by SukkaW** [comment]
Now we only have 3 failed test cases. We are nearly there!

**Comment by SukkaW** [comment]
The PR is now landed!

**Comment by JoshuaKGoldberg** [comment]
Whoohoo, thanks + congrats @SukkaW! 🥳

**Comment by merrywhether** [comment]
Awesome, glad I could kickstart this back up 👯

**Comment by xsjcTony** [comment]
Congrats!!!!!

**Comment by GaborTorma** [comment]
Awesome. Thanx your work! I waiting for the release.

**Pull request merged**